File: examples/dpop-client/typescript/dpop-client.ts
/**
 * DPoP Client Example — TypeScript
 *
 * Demonstrates how to:
 * 1. Generate an EC P-256 key pair
 * 2. Create a DPoP proof JWT
 * 3. Request a DPoP-bound access token from StrongDM ID
 * 4. Make an authenticated API call with a fresh DPoP proof
 *
 * Requirements: npm install jose
 * Run: npx tsx dpop-client.ts
 */

import * as jose from "jose";

const ISSUER = process.env.STRONGDM_ISSUER || "https://id.strongdm.ai";
const CLIENT_ID = process.env.STRONGDM_CLIENT_ID || "";
const CLIENT_SECRET = process.env.STRONGDM_CLIENT_SECRET || "";

/**
 * Create a DPoP proof JWT.
 *
 * @param privateKey - The EC private key to sign the proof with
 * @param publicJwk - The corresponding public JWK (embedded in proof header)
 * @param method - HTTP method (e.g., "POST", "GET")
 * @param url - Target URL (without query string)
 * @param accessToken - If provided, includes an ath (access token hash) claim
 */
async function createDPoPProof(
  privateKey: CryptoKey,
  publicJwk: jose.JWK,
  method: string,
  url: string,
  accessToken?: string
): Promise<string> {
  const payload: Record<string, unknown> = {
    jti: crypto.randomUUID(),
    htm: method,
    htu: url,
    iat: Math.floor(Date.now() / 1000),
  };

  // Include access token hash when proving possession for an API call
  if (accessToken) {
    const encoder = new TextEncoder();
    const hash = await crypto.subtle.digest(
      "SHA-256",
      encoder.encode(accessToken)
    );
    payload.ath = jose.base64url.encode(new Uint8Array(hash));
  }

  return new jose.SignJWT(payload)
    .setProtectedHeader({
      typ: "dpop+jwt",
      alg: "ES256",
      jwk: publicJwk,
    })
    .sign(privateKey);
}

/**
 * Request a DPoP-bound access token using client_credentials grant.
 */
async function requestDPoPToken(
  privateKey: CryptoKey,
  publicJwk: jose.JWK,
  scope: string = "openid"
): Promise<Record<string, unknown>> {
  const tokenUrl = `${ISSUER}/token`;
  const proof = await createDPoPProof(privateKey, publicJwk, "POST", tokenUrl);
  const credentials = btoa(`${CLIENT_ID}:${CLIENT_SECRET}`);

  const resp = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Authorization: `Basic ${credentials}`,
      DPoP: proof,
    },
    body: new URLSearchParams({
      grant_type: "client_credentials",
      scope,
    }),
  });

  // Handle DPoP nonce requirement
  if (resp.status === 400) {
    const body = await resp.json();
    if (body.error === "use_dpop_nonce") {
      const nonce = resp.headers.get("DPoP-Nonce");
      if (nonce) {
        console.log("Server requires DPoP nonce, retrying...");
        const retryProof = await createDPoPProof(
          privateKey,
          publicJwk,
          "POST",
          tokenUrl
        );
        const retryResp = await fetch(tokenUrl, {
          method: "POST",
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            Authorization: `Basic ${credentials}`,
            DPoP: retryProof,
          },
          body: new URLSearchParams({
            grant_type: "client_credentials",
            scope,
          }),
        });
        if (!retryResp.ok) {
          const errBody = await retryResp.text();
          throw new Error(`Token retry failed: HTTP ${retryResp.status}: ${errBody}`);
        }
        return retryResp.json();
      }
    }
    throw new Error(`Token request failed: ${JSON.stringify(body)}`);
  }

  if (!resp.ok) {
    const errBody = await resp.text();
    throw new Error(`Token request failed: HTTP ${resp.status}: ${errBody}`);
  }

  return resp.json();
}

/**
 * Make an API call with a DPoP-bound access token.
 */
async function callWithDPoP(
  url: string,
  accessToken: string,
  privateKey: CryptoKey,
  publicJwk: jose.JWK,
  method: string = "GET"
): Promise<Response> {
  const proof = await createDPoPProof(
    privateKey,
    publicJwk,
    method,
    url,
    accessToken
  );

  return fetch(url, {
    method,
    headers: {
      Authorization: `DPoP ${accessToken}`,
      DPoP: proof,
    },
  });
}

async function main() {
  if (!CLIENT_ID || !CLIENT_SECRET) {
    console.error(
      "Set STRONGDM_CLIENT_ID and STRONGDM_CLIENT_SECRET environment variables"
    );
    process.exit(1);
  }

  // Step 1: Generate a fresh EC P-256 key pair
  console.log("Generating EC P-256 key pair...");
  const { publicKey, privateKey } = await jose.generateKeyPair("ES256");
  const publicJwk = await jose.exportJWK(publicKey);
  const thumbprint = await jose.calculateJwkThumbprint(publicJwk, "sha256");
  console.log(`JWK thumbprint: ${thumbprint}`);

  // Step 2: Request a DPoP-bound token
  console.log(`\nRequesting DPoP token from ${ISSUER}/token...`);
  const tokenResponse = await requestDPoPToken(privateKey, publicJwk);

  const accessToken = tokenResponse.access_token as string;
  const tokenType = tokenResponse.token_type as string;
  const expiresIn = tokenResponse.expires_in as number;

  console.log(`Token type: ${tokenType}`);
  console.log(`Expires in: ${expiresIn}s`);
  console.log(`Token (first 40 chars): ${accessToken.slice(0, 40)}...`);

  // Step 3: Verify the token has cnf.jkt matching our key
  const claims = jose.decodeJwt(accessToken);
  const cnf = claims.cnf as { jkt?: string } | undefined;
  if (cnf?.jkt) {
    console.log(`\nToken cnf.jkt: ${cnf.jkt}`);
    console.log(`Our thumbprint: ${thumbprint}`);
    if (cnf.jkt === thumbprint) {
      console.log("Thumbprint matches — token is bound to our key");
    } else {
      console.log("WARNING: thumbprint mismatch");
    }
  }

  // Step 4: Example API call (uncomment and set your URL)
  // const apiUrl = "https://your-api.example.com/protected";
  // console.log(`\nCalling ${apiUrl} with DPoP proof...`);
  // const resp = await callWithDPoP(apiUrl, accessToken, privateKey, publicJwk);
  // console.log(`Response: HTTP ${resp.status}`);
  // console.log(await resp.text());
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: examples/dpop-client/python/dpop_client.py
"""
DPoP Client Example — Python

Demonstrates how to:
1. Generate an EC P-256 key pair
2. Create a DPoP proof JWT
3. Request a DPoP-bound access token from StrongDM ID
4. Make an authenticated API call with a fresh DPoP proof

Requirements: pip install jwcrypto httpx
"""

import json
import os
import sys
import time
import uuid
import hashlib
import base64

import httpx
from jwcrypto import jwk, jwt as jwcrypto_jwt


def generate_dpop_key() -> jwk.JWK:
    """Generate a fresh EC P-256 key pair for DPoP proofs."""
    return jwk.JWK.generate(kty="EC", crv="P-256")


def create_dpop_proof(
    key: jwk.JWK,
    method: str,
    url: str,
    access_token: str | None = None,
) -> str:
    """
    Create a DPoP proof JWT.

    Args:
        key: The EC key pair to sign the proof with.
        method: HTTP method (e.g., "POST", "GET").
        url: The target URL (without query string).
        access_token: If provided, includes an ath (access token hash) claim.

    Returns:
        The signed DPoP proof JWT string.
    """
    # Public key for the JWK header (no private material)
    pub = json.loads(key.export_public())

    header = {
        "typ": "dpop+jwt",
        "alg": "ES256",
        "jwk": pub,
    }

    payload = {
        "jti": str(uuid.uuid4()),
        "htm": method,
        "htu": url,
        "iat": int(time.time()),
    }

    # Include access token hash when proving possession for an API call
    if access_token is not None:
        token_hash = hashlib.sha256(access_token.encode()).digest()
        payload["ath"] = base64.urlsafe_b64encode(token_hash).rstrip(b"=").decode()

    token = jwcrypto_jwt.JWT(header=header, claims=payload)
    token.make_signed_token(key)
    return token.serialize()


def jwk_thumbprint(key: jwk.JWK) -> str:
    """Compute the RFC 7638 JWK thumbprint (base64url-encoded SHA-256)."""
    return key.thumbprint()


def request_dpop_token(
    issuer: str,
    client_id: str,
    client_secret: str,
    key: jwk.JWK,
    scope: str = "openid",
) -> dict:
    """
    Request a DPoP-bound access token using client_credentials grant.

    Returns the full token response dict.
    """
    token_url = f"{issuer}/token"

    # Create the DPoP proof for the token request
    proof = create_dpop_proof(key, "POST", token_url)

    auth_value = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()

    with httpx.Client(timeout=30) as client:
        resp = client.post(
            token_url,
            data={
                "grant_type": "client_credentials",
                "scope": scope,
            },
            headers={
                "Authorization": f"Basic {auth_value}",
                "DPoP": proof,
                "Content-Type": "application/x-www-form-urlencoded",
            },
        )

        # Handle DPoP nonce requirement — server may respond with
        # use_dpop_nonce error and a DPoP-Nonce header
        if resp.status_code == 400:
            body = resp.json()
            if body.get("error") == "use_dpop_nonce":
                nonce = resp.headers.get("DPoP-Nonce")
                if nonce:
                    print(f"Server requires DPoP nonce, retrying...")
                    proof = create_dpop_proof(key, "POST", token_url)
                    resp = client.post(
                        token_url,
                        data={
                            "grant_type": "client_credentials",
                            "scope": scope,
                        },
                        headers={
                            "Authorization": f"Basic {auth_value}",
                            "DPoP": proof,
                            "Content-Type": "application/x-www-form-urlencoded",
                        },
                    )

        if not resp.is_success:
            print(f"Token request failed: HTTP {resp.status_code}")
            print(resp.text)
            sys.exit(1)

        return resp.json()


def call_api_with_dpop(
    url: str,
    access_token: str,
    key: jwk.JWK,
    method: str = "GET",
) -> httpx.Response:
    """
    Make an API call with a DPoP-bound token.

    Each call needs a fresh DPoP proof bound to the specific method and URL.
    """
    proof = create_dpop_proof(key, method, url, access_token=access_token)

    with httpx.Client(timeout=30) as client:
        return client.request(
            method,
            url,
            headers={
                "Authorization": f"DPoP {access_token}",
                "DPoP": proof,
            },
        )


def main():
    issuer = os.environ.get("STRONGDM_ISSUER", "https://id.strongdm.ai")
    client_id = os.environ.get("STRONGDM_CLIENT_ID")
    client_secret = os.environ.get("STRONGDM_CLIENT_SECRET")

    if not client_id or not client_secret:
        print("Set STRONGDM_CLIENT_ID and STRONGDM_CLIENT_SECRET environment variables")
        sys.exit(1)

    # Step 1: Generate a fresh key pair
    print("Generating EC P-256 key pair...")
    key = generate_dpop_key()
    thumbprint = jwk_thumbprint(key)
    print(f"JWK thumbprint: {thumbprint}")

    # Step 2: Request a DPoP-bound token
    print(f"\nRequesting DPoP token from {issuer}/token...")
    token_response = request_dpop_token(issuer, client_id, client_secret, key)

    access_token = token_response["access_token"]
    token_type = token_response.get("token_type", "")
    expires_in = token_response.get("expires_in", 0)

    print(f"Token type: {token_type}")
    print(f"Expires in: {expires_in}s")
    print(f"Token (first 40 chars): {access_token[:40]}...")

    # Step 3: Verify the token has cnf.jkt matching our key
    # (Decode the JWT payload to check — don't verify signature here,
    # just inspect the claims)
    parts = access_token.split(".")
    if len(parts) == 3:
        payload_bytes = base64.urlsafe_b64decode(parts[1] + "==")
        claims = json.loads(payload_bytes)
        cnf = claims.get("cnf", {})
        token_jkt = cnf.get("jkt", "")
        print(f"\nToken cnf.jkt: {token_jkt}")
        print(f"Our thumbprint: {thumbprint}")
        if token_jkt == thumbprint:
            print("Thumbprint matches — token is bound to our key")
        else:
            print("WARNING: thumbprint mismatch")

    # Step 4: Example API call (uncomment and set your API URL)
    # api_url = "https://your-api.example.com/protected"
    # print(f"\nCalling {api_url} with DPoP proof...")
    # resp = call_api_with_dpop(api_url, access_token, key)
    # print(f"Response: HTTP {resp.status_code}")
    # print(resp.text)


if __name__ == "__main__":
    main()
